src/node.rs

```diff
@@ -357,6 +538,368 @@ impl AudioProcessor for OscillatorRenderer {
+        sample
+    }
+}
+
+#[derive(Debug)]
+struct TriangleRenderer {
+    frequency: f32,
+    sample_rate: f32,
+    incr_phase: f32,
+    mu: f32,
+    phase: f32,
+}
+
+impl TriangleRenderer {
+    fn new(frequency: f32, sample_rate: f32) -> Self {
+        let incr_phase = TABLE_LENGTH_F32 * frequency / sample_rate;
+        let mu = (incr_phase - incr_phase.round()).abs();
+        Self {
+            frequency,
+            sample_rate,
+            phase: 0.0,
+            incr_phase,
+            mu,
+        }
+    }
+
+    fn set_frequency(&mut self, frequency: f32) {
+        self.incr_phase = TABLE_LENGTH_F32 * frequency / self.sample_rate;
+        self.mu = (self.incr_phase - self.incr_phase.round()).abs();
+        self.frequency = frequency;
+    }
+}
+
+impl PolyBlep for TriangleRenderer {
+    fn incr_phase(&self) -> f32 {
+        self.incr_phase
+    }
+}
+
+impl Ticker for TriangleRenderer {
+    fn tick(&mut self) -> f32 {
+        let idx = (self.phase + self.incr_phase) as usize;
+        let inf_idx = idx % TABLE_LENGTH_USIZE;
+        let sup_idx = (idx + 1) % TABLE_LENGTH_USIZE;
+
+        let mut sample = TRIANGLETABLE[inf_idx] * (1. - self.mu) + TRIANGLETABLE[sup_idx] * self.mu;
+
+        let norm_phase = self.phase / TABLE_LENGTH_F32;
+        sample -= self.poly_blep(norm_phase);
+        self.phase = (self.phase + self.incr_phase) % TABLE_LENGTH_F32;
+
+        sample
+    }
+}
+
+#[derive(Debug)]
+struct SquareRenderer {
+    frequency: f32,
+    sample_rate: f32,
+    incr_phase: f32,
+    mu: f32,
+    phase: f32,
+}
+
+impl SquareRenderer {
+    fn new(frequency: f32, sample_rate: f32) -> Self {
+        let incr_phase = (TABLE_LENGTH_F32 / sample_rate) * frequency;
+        let mu = (incr_phase - incr_phase.round()).abs();
+        Self {
+            frequency,
+            sample_rate,
+            phase: 0.0,
+            incr_phase,
+            mu,
+        }
+    }
+
+    fn set_frequency(&mut self, frequency: f32) {
+        self.incr_phase = (TABLE_LENGTH_F32 / self.sample_rate) * frequency;
+        self.mu = (self.incr_phase - self.incr_phase.round()).abs();
+        self.frequency = frequency;
+    }
+}
+
+impl PolyBlep for SquareRenderer {
+    fn incr_phase(&self) -> f32 {
+        self.incr_phase
+    }
+}
+
+impl Ticker for SquareRenderer {
+    fn tick(&mut self) -> f32 {
+        let idx = (self.phase + self.incr_phase) as usize;
+        let inf_idx = idx % TABLE_LENGTH_USIZE;
+        let sup_idx = (idx + 1) % TABLE_LENGTH_USIZE;
+
+        let mut sample = SQUARETABLE[inf_idx] * (1. - self.mu) + SQUARETABLE[sup_idx] * self.mu;
+
+        let norm_phase = self.phase / TABLE_LENGTH_F32;
+        sample -= self.poly_blep(norm_phase);
+        self.phase = (self.phase + self.incr_phase) % TABLE_LENGTH_F32;
+        sample
+    }
+}
+
+struct CustomRenderer {
+    frequency: f32,
+    sample_rate: f32,
+    cplxs: Vec<(f32, f32)>,
```

Comment:
There may be a performance improvement by packing these multiple Vecs as a single Vec of tuples. But let's see what profiling reports